src/language/lexer.rs

```diff
@@ -2,26 +2,26 @@
 use std::str::CharIndices;
 use std::iter::Peekable;
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq)]
 pub enum Token<'a> {
-  Eof,
-  Bang,
-  Dollar,
-  LeftParen,
-  RightParen,
-  Spread,
-  Colon,
-  Equals,
-  At,
-  LeftBracket,
-  RightBracket,
-  LeftBrace,
-  RightBrace,
-  Pipe,
-  Name(&'a str),
-  Int(&'a str),
-  Float(&'a str),
-  String(&'a str),
+  Eof(usize, usize),
+  Bang(usize, usize),
+  Dollar(usize, usize),
+  LeftParen(usize, usize),
+  RightParen(usize, usize),
+  Spread(usize, usize),
+  Colon(usize, usize),
+  Equals(usize, usize),
+  At(usize, usize),
+  LeftBracket(usize, usize),
+  RightBracket(usize, usize),
+  LeftBrace(usize, usize),
+  RightBrace(usize, usize),
+  Pipe(usize, usize),
+  Name(usize, usize, &'a str),
+  Int(usize, usize, &'a str),
+  Float(usize, usize, &'a str),
```

Comment:
Location information at the end sounds like a great idea, but as for punctuators it isn't so much a matter of performance to me, but rather a matter of I just really like enums 😊

Type safety and pattern exhaustion are super powerful tools in Rust (coming from a JavaScript/Haskell background), so we should really try to use them wherever possible. I agree it doesn't look the prettiest, but that's what macros/compiler plugins are for 😉

The iron web framework went this way with headers. For example a `Content-Type` header looks like: `ContentType(Mime(TopLevel::Text, SubLevel::Html, vec![(Attr::Charset, Value::Utf8)]))`. There justification (which I agree with even if I hate writing it) can be found [here](http://ironframework.io/doc/iron/headers/index.html#why-typed).

To further alleviate ugliness, we could also do what the [standard library does with `Result`](http://doc.rust-lang.org/std/prelude/index.html) in that we write `pub use Puncuator::*`. This changes `Token::Punctuator(Puncuator::Spread, _, _)` to just `Token::Puncuator(Spread, _, _)` a much cleaner and agreeable syntax.

Thoughts?